dialogue-core/src/main/java/com/palantir/dialogue/core/DialogueChannel.java

```diff
@@ -0,0 +1,276 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.Ticker;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Sets;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.errorprone.annotations.CheckReturnValue;
+import com.palantir.conjure.java.client.config.ClientConfiguration;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.random.SafeThreadLocalRandom;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+public final class DialogueChannel implements Channel {
+    private Map<String, LimitedChannel> limitedChannelByUri = new ConcurrentHashMap<>();
+    private final AtomicReference<LimitedChannel> nodeSelectionStrategy = new AtomicReference<>();
+
+    private final ClientConfiguration clientConfiguration;
+    private final ChannelFactory channelFactory;
+    private final Channel delegate;
+    private final DialogueClientMetrics clientMetrics;
+    private final Ticker clock;
+    private final Random random;
+
+    // TODO(forozco): you really want a refreshable of uri separate from the client config
+    private DialogueChannel(
+            ClientConfiguration clientConfiguration,
+            ChannelFactory channelFactory,
+            Ticker clock,
+            Random random,
+            Supplier<ScheduledExecutorService> scheduler) {
+        this.clientConfiguration = clientConfiguration;
+        this.channelFactory = channelFactory;
+        clientMetrics = DialogueClientMetrics.of(clientConfiguration.taggedMetricRegistry());
+        this.clock = clock;
+        this.random = random;
+        updateUris(clientConfiguration.uris());
+        this.delegate = wrap(
+                new SupplierChannel(nodeSelectionStrategy::get), clientConfiguration, scheduler, random, clientMetrics);
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        return delegate.execute(endpoint, request);
+    }
+
+    public void updateUris(List<String> uris) {
```

Comment:
Considered exposing an update method that would recreate the entire client if the client configuration wasn't identical (except for uris) but it seemed like more faff than it was worth